Feature/aws dynamic account

nabuskey opened the pull request:
We prefer small, well tested pull requests.

Please refer to [Contributing to Spinnaker](https://spinnaker.io/community/contributing/).

When filling out a pull request, please consider the following:

* Follow the commit message conventions [found here](https://spinnaker.io/community/contributing/submitting/).
* Provide a descriptive summary for your changes.
* If it fixes a bug or resolves a feature request, be sure to link to that issue.
* Add inline code comments to changes that might not be obvious.
* Squash your commits as you keep adding changes.
* Add a comment to @spinnaker/reviewers for review if your issue has been outstanding for more than 3 days.

Note that we are unlikely to accept pull requests that add features without prior discussion. The best way to propose a feature is to open an issue first and discuss your ideas there before implementing them.

nabuskey committed:
WIP

aravindmd committed:
feat(aws): Support cross zone load balancing for NLB (#4862)

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

spinnakerbot committed:
chore(dependencies): Autobump korkVersion (#4975)

Co-authored-by: root <root@af3522b9b644>

ncknt commented on clouddriver-api/src/main/java/com/netflix/spinnaker/clouddriver/security/CredentialsProvider.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.Set;
+
+public interface CredentialsProvider<T extends AccountCredentials<?>> {
```
I'd remove, this does not seem like a useful abstraction.

nabuskey replied:
Removed CredentialsProvider and AmazonCredentialsProvider.

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/controllers/AmazonClusterController.groovy:
```diff
@@ -31,7 +31,7 @@ import org.springframework.web.bind.annotation.*
 
   @Autowired
-  AccountCredentialsProvider accountCredentialsProvider
+  AmazonCredentialProvider<NetflixAmazonCredentials> accountCredentialsProvider
```
just use `CredentialsRepository<NetflixAmazonCredentials>`

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/AwsInfrastructureProvider.groovy:
```diff
@@ -23,11 +23,12 @@ import com.netflix.spinnaker.clouddriver.aws.AmazonCloudProvider
 import static com.netflix.spinnaker.clouddriver.cache.SearchableProvider.SearchableResource
 
-class AwsInfrastructureProvider extends AgentSchedulerAware implements SearchableProvider {
+class AwsInfrastructureProvider extends AgentSchedulerAware, BaseProvider implements SearchableProvider {
```
- Remove `AgentSchedulerAware` (it's in `BaseProvider`)
- Remove `agents` property as well.

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/AwsCleanupProvider.groovy:
```diff
@@ -19,8 +19,9 @@ package com.netflix.spinnaker.clouddriver.aws.provider
+import com.netflix.spinnaker.clouddriver.security.BaseProvider
 
-class AwsCleanupProvider extends AgentSchedulerAware implements Provider {
+class AwsCleanupProvider extends AgentSchedulerAware, BaseProvider implements Provider {
```
Remove `AgentSchedulerAware` (it's in `BaseProvider`)

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/AwsCleanupProvider.groovy:
```diff
@@ -19,8 +19,9 @@ package com.netflix.spinnaker.clouddriver.aws.provider
+class AwsCleanupProvider extends AgentSchedulerAware, BaseProvider implements Provider {
   public static final String PROVIDER_NAME = AwsCleanupProvider.name
 
   private final Collection<Agent> agents
```
Remove `agents` property

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/AwsProvider.groovy:
```diff
@@ -19,23 +19,24 @@ package com.netflix.spinnaker.clouddriver.aws.provider
 import static com.netflix.spinnaker.clouddriver.core.provider.agent.Namespace.*
 
-class AwsProvider extends AgentSchedulerAware implements SearchableProvider, EurekaAwareProvider {
+class AwsProvider extends AgentSchedulerAware, BaseProvider implements SearchableProvider, EurekaAwareProvider {
```
- Remove `AgentSchedulerAware` (it's in `BaseProvider`)
- Remove `agents` property as well.

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/AwsProvider.groovy:
```diff
@@ -58,7 +59,7 @@ class AwsProvider extends AgentSchedulerAware implements SearchableProvider, Eur
   private Collection<HealthProvidingCachingAgent> healthAgents
 
-  AwsProvider(AccountCredentialsRepository accountCredentialsRepository, Collection<Agent> agents) {
+  AwsProvider(CredentialsRepository<? extends NetflixAmazonCredentials> accountCredentialsRepository, Collection<Agent> agents) {
```
With the change above, it means `agents` aren't passed when instantiating the provider.

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonInstanceTypeCachingAgent.java:
```diff
@@ -102,7 +106,7 @@ public CacheResult loadData(ProviderCache providerCache) {
       Set<String> matchingAccounts =
           accountCredentialsRepository.getAll().stream()
-              .filter(AmazonCredentials.class::isInstance)
+              .filter(Objects::nonNull)
```
I don't believe credentials can be null.

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentialsInitializer.groovy:
```diff
@@ -47,28 +61,77 @@ class AmazonCredentialsInitializer {
   }
 
+  @Bean
+  @DependsOn("amazonCredentialsLifecycleHandler")
```
Is this needed?

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentialsLifecycleHandler.java:
```diff
@@ -0,0 +1,189 @@
+            registry,
+            eddaTimeoutConfig,
+            this.awsInfraRegions);
+    if (awsInfrastructureProvider.getAgentScheduler() != null) {
```
Replace with `awsProvider.removeAgentsForAccounts(Collections.singleton(credentials.getName()))` for each provider being updated

ncknt replied:
(See KubernetesLifecycleHandler for example)

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentialsLifecycleHandler.java:
```diff
@@ -0,0 +1,189 @@
+            amazonS3DataProvider,
+            publicRegions);
+    if (awsProvider.getAgentScheduler() != null) {
+      ProviderUtils.rescheduleAgents(awsProvider, buildResult.agents);
```
This shouldn't be needed. See `BaseProvider.addAgents()`.

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentialsLifecycleHandler.java:
```diff
@@ -0,0 +1,189 @@
+/*
```
Should it we start the `java/` directory?

ncknt commented on clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaAgentProvider.java:
```diff
@@ -57,8 +58,8 @@ public boolean supports(String providerName) {
-    accountCredentialsProvider.getAll().stream()
-        .filter(c -> c instanceof NetflixAmazonCredentials)
+    credentialsRepository.getAll().stream()
+        .filter(Objects::nonNull)
```
I don't believe credentials can be null

ncknt commented on clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/security/EcsCredentialsInitializer.java:
```diff
@@ -40,20 +41,24 @@ public ECSCredentialsConfig ecsCredentialsConfig() {
-      CredentialsLoader<? extends NetflixAmazonCredentials> credentialsLoader,
-      ECSCredentialsConfig credentialsConfig,
-      AccountCredentialsRepository accountCredentialsRepository)
+      CredentialsRepository<NetflixAmazonCredentials> accountCredentialsRepository,
```
nitpick but can we call it `credentialsRepository` instead of `accountCredentialsRepository` - we're trying to be clear with the terms used.

ncknt commented on clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/security/EcsCredentialsInitializer.java:
```diff
@@ -40,20 +41,24 @@ public ECSCredentialsConfig ecsCredentialsConfig() {
 
   @Bean
-  @DependsOn("netflixAmazonCredentials")
+  @DependsOn("amazonCredentialsLoader")
```
I don't think it's needed.

ncknt commented on clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/security/EcsCredentialsInitializer.java:
```diff
@@ -40,20 +41,24 @@ public ECSCredentialsConfig ecsCredentialsConfig() {
+      AmazonCredentialsParser<CredentialsConfig.Account, ? extends NetflixAmazonCredentials>
+          amazonCredentialsParser,
+      ECSCredentialsConfig ecsCredentialsConfig)
       throws Throwable {
```
`Throwable` is a little too much :) Do we even need an exception declared here?

ncknt commented on clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/security/EcsCredentialsInitializer.java:
```diff
@@ -78,11 +83,12 @@ public ECSCredentialsConfig ecsCredentialsConfig() {
             NetflixECSCredentials ecsCredentials =
                 new NetflixAssumeRoleEcsCredentials(
-                    (NetflixAssumeRoleAmazonCredentials) credentialsLoader.load(ecsCopy).get(0),
+                    (NetflixAssumeRoleAmazonCredentials)
```
The loop needs to be changed to not use `AccountCredentials` but `NetflixAmazonCredentials` without cast.

jeyrschabu committed:
enhancement(aws): Updating launch template rollout flag (#4977)

- updated exclusion to be similar to inclusion in format
- added regex support to include apps with similar naming

plumpy committed:
chore(kotlin): use the JDK 11 kotlin libs (#4978)

I missed this in commit #4919, which switched us to targeting Java 11.

spinnakerbot committed:
chore(dependencies): Autobump spinnakerGradleVersion (#4979)

Co-authored-by: root <root@6a7b3894974d>

spinnakerbot committed:
chore(dependencies): Autobump korkVersion (#4980)

Co-authored-by: root <root@d7e7980769a5>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

spinnakerbot committed:
chore(dependencies): Autobump korkVersion (#4981)

Co-authored-by: root <root@807cc584393a>

spinnakerbot committed:
chore(dependencies): Autobump fiatVersion (#4982)

Co-authored-by: root <root@b272e1bb32f7>

spinnakerbot committed:
chore(dependencies): Autobump korkVersion (#4983)

Co-authored-by: root <root@a88a9ef8bf72>

ajordens committed:
fix(titus): Improve handling for a non-existent Titus Job (#4947)

Rather than fail the action, no-op the delete and return an empty `DeleteServerGroupEvent`.

rajkdale committed:
fix(Oracle): Fixed version support for deploying artifacts from Oracle object storage (#4949)

* fix(Oracle): Fixed version support deploy artifacts in Oracle objects storage

* Updated pound to versionIndex

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

spinnakerbot committed:
chore(dependencies): Autobump spinnakerGradleVersion (#4986)

Co-authored-by: root <root@5900c8b382c3>

plumpy committed:
fix(gce): remove the ONLY_DOWN deprecated AutoscalingMode (#4985)

* feat(gce): remove the deprecated ONLY_UP autoscaling mode

ONLY_SCALE_OUT is the replacement. We'll transform any ONLY_UP we see into ONLY_SCALE_OUT as soon as we find it so that we can pretend ONLY_UP doesn't even exist.

* fix(gce): remove the ONLY_DOWN deprecated AutoscalingMode

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

fieldju committed:
fix(aws): use the Spinnaker region provider with the sts client builder, so that it defaults to the default region, when no region is defined (#4989)

german-muzquiz committed:
fix(cf): Build sg names using existing methods (#4987)

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/AllowLaunchAtomicOperation.groovy:
```diff
@@ -51,14 +52,14 @@ class AllowLaunchAtomicOperation implements AtomicOperation<ResolvedAmiResult> {
 
     def sourceCredentials = description.credentials
-    def targetCredentials = accountCredentialsProvider.getCredentials(description.targetAccount) as NetflixAmazonCredentials
+    def targetCredentials = credentialsRepository.getOne(description.targetAccount) as NetflixAmazonCredentials
```
You should be able to drop `as NetflixAmazonCredentials`

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/CopyLastAsgAtomicOperation.groovy:
```diff
@@ -90,7 +92,7 @@ class CopyLastAsgAtomicOperation implements AtomicOperation<DeploymentResult> {
       if (description.source.account && description.source.region && description.source.asgName) {
         sourceRegion = description.source.region
-        sourceAsgCredentials = accountCredentialsProvider.getCredentials(description.source.account) as NetflixAmazonCredentials
+        sourceAsgCredentials = credentialsRepository.getOne(description.source.account) as NetflixAmazonCredentials
```
You should be able to drop `as NetflixAmazonCredentials`

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/validators/AllowLaunchDescriptionValidator.groovy:
```diff
@@ -38,7 +39,7 @@ class AllowLaunchDescriptionValidator extends DescriptionValidator<AllowLaunchDe
     if (!description.targetAccount) {
       errors.rejectValue("targetAccount", "allowLaunchDescription.targetAccount.empty")
-    } else if (!accountCredentialsProvider.all.collect { it.name }.contains(description.targetAccount)) {
+    } else if (!credentialsRepository.getAll().collect { it.name }.contains(description.targetAccount)) {
```
`credentialsRepository.getOne(description.targetAccount) == null`

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/validators/BasicAmazonDeployDescriptionValidator.groovy:
```diff
@@ -41,7 +42,7 @@ class BasicAmazonDeployDescriptionValidator extends AmazonDescriptionValidationS
     } else {
-      credentials = accountCredentialsProvider.getCredentials(description?.credentials?.name)
+      credentials = credentialsRepository.getOne(description?.credentials?.name)
       if (!(credentials instanceof AmazonCredentials)) {
```
You can just check `credentials == null` since `NetflixAmazonCredentials` extends `AmazonCredentials`

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/validators/ModifyAsgLaunchConfigurationDescriptionValidator.groovy:
```diff
@@ -40,7 +41,7 @@ class ModifyAsgLaunchConfigurationDescriptionValidator extends AmazonDescription
     } else {
-      def credentials = accountCredentialsProvider.getCredentials(description?.credentials?.name)
+      def credentials = credentialsRepository.getOne(description?.credentials?.name)
       if (!(credentials instanceof AmazonCredentials)) {
```
You can just check `credentials == null` since `NetflixAmazonCredentials` extends `AmazonCredentials`

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/health/AmazonHealthIndicator.groovy:
```diff
@@ -80,7 +79,7 @@ class AmazonHealthIndicator implements HealthIndicator {
   void checkHealth() {
     try {
-      Set<NetflixAmazonCredentials> amazonCredentials = accountCredentialsProvider.all.findAll {
+      Set<NetflixAmazonCredentials> amazonCredentials = credentialsRepository.getAll().findAll {
```
`credentialsRepository.getAll().forEach(....)` the test for `instanceof` is always true

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/AwsCleanupProvider.groovy:
```diff
@@ -19,23 +19,15 @@ package com.netflix.spinnaker.clouddriver.aws.provider
 
-class AwsCleanupProvider extends AgentSchedulerAware implements Provider {
+@Component
+class AwsCleanupProvider extends BaseProvider implements Provider {
```
`BaseProvider` already implements `Provider`.

ncknt replied:
Also it probably shouldn't be created as a `Component` but in the configuration class as a Bean (as currently) since it is controlled by `aws.enabled`

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/AwsInfrastructureProvider.groovy:
```diff
@@ -17,39 +17,27 @@
 
-class AwsInfrastructureProvider extends AgentSchedulerAware implements SearchableProvider {
+@Component
+class AwsInfrastructureProvider extends BaseProvider implements SearchableProvider {
```
Same comment about not creating as `@component` but as a bean in the configuration class

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/AwsProvider.groovy:
```diff
@@ -133,14 +130,14 @@ class AwsProvider extends AgentSchedulerAware implements SearchableProvider, Eur
     if (allowMultipleEurekaPerAccount) {
-      def credentialName = accountCredentialsRepository.all.find {
+      def credentialName = credentialsRepository.all.find {
         it instanceof NetflixAmazonCredentials && it.accountId == accountId && it.name == eurekaAccountName
```
unnecessary `instanceof` check

ncknt commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/AwsProvider.groovy:
```diff
@@ -133,14 +130,14 @@ class AwsProvider extends AgentSchedulerAware implements SearchableProvider, Eur
     }
-    return accountCredentialsRepository.all.find {
+    return credentialsRepository.all.find {
       it instanceof NetflixAmazonCredentials && it.accountId == accountId
```
unnecessary `instanceof` check

zachsmith1 committed:
fix(provider/cf): fix regression caused by clusterSummaries (#4991)

Co-authored-by: Kevin Woo <kevinawoo@gmail.com>

robzienert committed:
fix(titus): Store application ref for server groups and instances (#4992)

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

robzienert committed:
Revert "fix(titus): Store application ref for server groups and instances (#4992)" (#4993)

This reverts commit 7f81bddf75a20782c6bc8257db180a05370799c1.

plumpy committed:
chore(kork): update kork dependency (now w/ moniker) (#4995)

ezimanyi committed:
chore(codeowners): Add german-muzquiz to kubernetes codeowners (#4939)

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

ajordens committed:
feat(titus): Converted a few more Titus operations to sagas (#4990)

* `detachInstances`
* `resizeServerGroup`
* `terminateInstances`

I opted to use Titus terminology when naming the actions (Tasks vs Instances and
Job vs Server Group).

tompiscitell committed:
feat(clouddriver): Add support for Redis SSL (#4962)

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

aravindmd committed:
fix(titus): Update error handling (#4938)

* fix(titus): Update error handling

* minor update

* update to use spinnaker exception

* fixup

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

emagana committed:
feat(azure): Enable Azure from cloud providers (#5000)

SIG Azure has be created
Teams are interested in getting the support back

robzienert committed:
perf(cats/sql): Introduce V2 SQL cache schema (#4998)

* fix(cats/sql): Reintroduce fix to Clouddriver SQL Cache failing due to id or agent being too long to be stored.

Revert "fix(cats/sql): Reverting changes to SQL Cache due to seeing a major increase in CPU usage and deadlocking. (#4811)"

This reverts commit 618921b0d0f8f9f1aed30e16a8f44abff6348acd.

* perf(cats/sql): Change relationship table index with rel_agent_hash to include rel_type.

Including the rel_type in the the index will improve the query where rel_agent_hash and rel_type is given. There's also a query to get all entries with a given rel_agent_hash, this index still works for that query since the rel_agent_hash is the first column.

* perf(cats/sql): Switch SqlCache hashing function from SHA256 to Murmur3.

Moving towards using a faster hashing function since we're spending a lot of CPU time hashing at the moment. Furthermore, the values we are hashing do not need to be hashed by a cryptographic hash function so SHA256 is overkill.

* fix(cats): Fix regressions in nullable identifiers

Co-authored-by: Matt Santa <mattsanta@google.com>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

Misdre committed:
feat(aws): update awscli version from 1.16.258 to 1.18.152 and s3cmd from 2.0.1 to 2.0.2 (#4976)

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

plumpy committed:
chore(kt): switch back to kotlin-stdlib-jdk8 (#5001)

spinnakerbot committed:
chore(dependencies): Autobump korkVersion (#5002)

Co-authored-by: root <root@2897678eb22e>

jeyrschabu committed:
fix(aws): Fixing NPE while retrieving imageId in launch templates (#5003)

- updated to get image id from launch template data
- made lookup null-safe

spinnakerbot committed:
chore(dependencies): Autobump spinnakerGradleVersion (#5006)

Co-authored-by: root <root@877703f80d77>

allisaurus committed:
feat(test): extend ECS CreateServerGroup integ tests (#4972)

* feat(test): extend ECS CreateServerGroup integ tests

* remove unused logger

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

robzienert committed:
Revert "perf(cats/sql): Introduce V2 SQL cache schema (#4998)" (#5008)

This reverts commit 5dd86f59b690df65e192f1dfbf655ad3fa04b2c8.

vasiliy-briginets committed:
feat(yandex): add a new cloud provider (#4974)

* feat(yandex): Support provider Yandex.Cloud

* fix(yandex): Upgrade Yandex.Cloud SDK version and refactoring

update kork dependency (now w/ moniker)

Co-authored-by: Vasiliy Briginets <0x40@yandex-team.ru>
Co-authored-by: Sergey Galyamichev <gsv@smtp.ru>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>
